### ISSUE TITLE
Generator having trouble with inheritance

### ISSUE THREAD

**Issue by JCKodel** [simolus3/moor#2462]
I have a base table that every table in my project inherits:

```dart
import 'package:flutter/foundation.dart';

import 'package:drift/drift.dart';

import 'database.dart';

abstract class SyncTable extends Table {
  DateTimeColumn get updateDate => dateTime().clientDefault(() => DateTime.now())();
  BoolColumn get synced => boolean().clientDefault(() => false)();
  BoolColumn get deleted => boolean().clientDefault(() => false)();

  @protected
  @mustCallSuper
  Future<void> onCreate(Database db, Migrator m) async {
    await createIndex(m, {updateDate}, true);
    await createIndex(m, {synced});
    await createIndex(m, {deleted});

    final entityName = (this as TableInfo).entityName;

    await m.createTrigger(
      Trigger(
        "CREATE TRIGGER ${entityName}_UpdateTrigger "
            "AFTER UPDATE ON ${entityName} FOR EACH ROW BEGIN "
            "UPDATE ${entityName} SET "
            "updateDate = STRFTIME('%Y-%m-%dT%H:%M:%fZ', 'now'),"
            "synced = 0 "
            "WHERE rowid = NEW.rowid;"
            "END;",
        "${entityName}_UpdateTrigger",
      ),
    );
  }

  @protected
  @mustCallSuper
  Future<void> onUpgrade(Database db, Migrator m, int from, int to) async {}

  @protected
  Future<void> createIndex(Migrator m, Set<Column> columns, [bool? descending]) async {
    final tableInfo = this as TableInfo;

    final indexName = "IX_${tableInfo.entityName}_"
        "${columns.map((c) => c.name).join("_")}";

    await m.createIndex(
      Index(
        indexName,
        "CREATE INDEX ${indexName} ON \"${tableInfo.entityName}\"("
        "${columns.map((c) => "${c.escapedName}"
            "${descending == true ? " DESC" : ""}").join(", ")}"
        ");",
      ),
    );
  }
}
```

Then, inheriting from it as:

```dart
import 'package:drift/drift.dart';

import '../../../k_authentication/k_authentication.dart';
import '../../shared/data/database.dart';
import '../../shared/data/sync_table.dart';

import 'users.dart';

@DataClassName("UserLogin")
class UserLogins extends SyncTable {
  TextColumn get id => text()();
  TextColumn get userId => text().references(Users, #id)();
  TextColumn get authenticationProvider => textEnum<AuthenticationProvider>()();

  @override
  Set<Column> get primaryKey => {id};

  @override
  Future<void> onCreate(Database db, Migrator m) async {
    await super.onCreate(db, m);
    await createIndex(m, {userId});
  }

  // Not needed for now
  // Future<void> onUpgrade(Database db, Migrator m, int from, int to) async {}
}
```

So far, no issues at all, project compiles an runs fine.

Now, when I try to create a .drift file using some table:

```sql
import 'capillary_schedules.dart';
import 'capillary_schedule_steps.dart';
import 'capillary_schedule_step_images.dart';

getDraftCapillarySchedule(REQUIRED :userId AS TEXT):
  SELECT cs.**,
  LIST(
    SELECT css.* 
    FROM CapillaryScheduleSteps AS css 
    WHERE css.capillaryScheduleId = cs.id
  ) AS steps
  FROM CapillarySchedules AS cs
  WHERE cs.deleted = 0 AND cs.userId = :userId AND cs.status = 'draft';
```

After that, my DAO starts to complain about a missing class on name_dao.g.dart:

```dart
part of 'hair_care_dao.dart';

// ignore_for_file: type=lint
mixin _$HairCareDaoMixin on DatabaseAccessor<Database> {
  $UsersTable get users => attachedDatabase.users;
  $CapillarySchedulesTable get capillarySchedules =>
      attachedDatabase.capillarySchedules;
  $CapillaryScheduleStepsTable get capillaryScheduleSteps =>
      attachedDatabase.capillaryScheduleSteps;
  $AttachedImagesTable get attachedImages => attachedDatabase.attachedImages;
  $CapillaryScheduleStepImagesTable get capillaryScheduleStepImages =>
      attachedDatabase.capillaryScheduleStepImages;
  $UserLoginsTable get userLogins => attachedDatabase.userLogins;
  $UserDevicesTable get userDevices => attachedDatabase.userDevices;
  $SyncTableTable get syncTable => attachedDatabase.syncTable;
  Selectable<GetDraftCapillaryScheduleResult> getDraftCapillarySchedule(

...

```

That `$SyncTableTable` should not exist (it is created ONLY when a .drift file has contents. If I just comment it out, it works fine).

**Comment by simolus3** [comment]
I get that this is annoying, but it's working as intended. Drift allows table classes to be abstract, so `SyncTable` is a valid table on its own. When you import a Dart file into a drift file, we assume that all tables defined in the Dart file exist, so this is what you get.

To work around this, you could make `SyncTable` a mixin `on Table` I guess.

**Comment by JCKodel** [comment]
Mixin won't work because we need to inherit from Table to be able to use `text()()`.

> The class 'SyncTable' can't be declared a mixin because it extends a class other than 'Object'.

> The class 'SyncTable' can't be used as a mixin because it extends a class other than 'Object'.

> Drift allows table classes to be abstract, so SyncTable is a valid table

But I didn't mark the SyncTable class with `@UseRowClass` or `@DataClass` O.o

**Comment by JCKodel** [comment]
After trying to 

```dart
abstract mixin class SyncTable {
  DateTimeColumn get updateDate =>
      ColumnBuilder<DateTime>().clientDefault(() => DateTime.now())();

  BoolColumn get synced => ColumnBuilder<bool>().clientDefault(() => false)();

  BoolColumn get deleted => ColumnBuilder<bool>().clientDefault(() => false)();
}
```

This is the result:

```console
[INFO] Generating build script completed, took 182ms
[INFO] Reading cached asset graph completed, took 92ms
[INFO] Checking for updates since last build completed, took 655ms
[WARNING] drift_dev on lib/features/hair_care/data/capillary_schedules.dart:
Could not analyze DriftElementId(package:maismulher/features/hair_care/data/capillary_schedules.dart, CapillarySchedules)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/authentication/data/users.dart:
Could not analyze DriftElementId(package:maismulher/features/authentication/data/users.dart, Users)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/authentication/data/user_logins.dart:
Could not analyze DriftElementId(package:maismulher/features/authentication/data/user_logins.dart, UserLogins)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/hair_care/data/capillary_schedule_step_images.dart:
Could not analyze DriftElementId(package:maismulher/features/hair_care/data/capillary_schedule_step_images.dart, CapillaryScheduleStepImages)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/authentication/data/user_devices.dart:
Could not analyze DriftElementId(package:maismulher/features/authentication/data/user_devices.dart, UserDevices)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/hair_care/data/capillary_schedule_steps.dart:
Could not analyze DriftElementId(package:maismulher/features/hair_care/data/capillary_schedule_steps.dart, CapillaryScheduleSteps)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/hair_care/data/hair_care_dao.dart:
Could not analze DriftElementId(package:maismulher/features/hair_care/data/capillary_schedules.dart, CapillarySchedules)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/hair_care/data/hair_care_dao.dart:
Could not analze DriftElementId(package:maismulher/features/hair_care/data/capillary_schedule_steps.dart, CapillaryScheduleSteps)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/hair_care/data/hair_care_dao.dart:
Could not analze DriftElementId(package:maismulher/features/hair_care/data/capillary_schedule_step_images.dart, CapillaryScheduleStepImages)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/hair_care/data/hair_care_dao.dart:
Could not analze DriftElementId(package:maismulher/features/storage/data/attached_images.dart, AttachedImages)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/hair_care/data/hair_care_dao.dart:
line 22, column 7 of package:maismulher/features/hair_care/data/hair_care_dao.dart: The referenced element could not be analyzed due to a bug in drift.
   ╷
22 │ class HairCareDao extends DatabaseAccessor<Database> with _$HairCareDaoMixin {
   │       ^^^^^^^^^^^
   ╵
[WARNING] drift_dev on lib/features/hair_care/data/hair_care_dao.dart:
line 22, column 7 of package:maismulher/features/hair_care/data/hair_care_dao.dart: The referenced element could not be analyzed due to a bug in drift.
   ╷
22 │ class HairCareDao extends DatabaseAccessor<Database> with _$HairCareDaoMixin {
   │       ^^^^^^^^^^^
   ╵
[WARNING] drift_dev on lib/features/hair_care/data/hair_care_dao.dart:
line 22, column 7 of package:maismulher/features/hair_care/data/hair_care_dao.dart: The referenced element could not be analyzed due to a bug in drift.
   ╷
22 │ class HairCareDao extends DatabaseAccessor<Database> with _$HairCareDaoMixin {
   │       ^^^^^^^^^^^
   ╵
[WARNING] drift_dev on lib/features/hair_care/data/hair_care_dao.dart:
line 22, column 7 of package:maismulher/features/hair_care/data/hair_care_dao.dart: The referenced element could not be analyzed due to a bug in drift.
   ╷
22 │ class HairCareDao extends DatabaseAccessor<Database> with _$HairCareDaoMixin {
   │       ^^^^^^^^^^^
   ╵
[WARNING] drift_dev on lib/features/shared/data/database.dart:
Could not analze DriftElementId(package:maismulher/features/authentication/data/users.dart, Users)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/shared/data/database.dart:
Could not analze DriftElementId(package:maismulher/features/authentication/data/user_logins.dart, UserLogins)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/shared/data/database.dart:
Could not analze DriftElementId(package:maismulher/features/authentication/data/user_devices.dart, UserDevices)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/shared/data/database.dart:
Could not analze DriftElementId(package:maismulher/features/hair_care/data/capillary_schedules.dart, CapillarySchedules)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/shared/data/database.dart:
Could not analze DriftElementId(package:maismulher/features/hair_care/data/capillary_schedule_steps.dart, CapillaryScheduleSteps)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/shared/data/database.dart:
Could not analze DriftElementId(package:maismulher/features/hair_care/data/capillary_schedule_step_images.dart, CapillaryScheduleStepImages)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/shared/data/database.dart:
Could not analze DriftElementId(package:maismulher/features/storage/data/attached_images.dart, AttachedImages)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/shared/data/database.dart:
Could not analze DriftElementId(package:maismulher/features/hair_care/data/capillary_schedules.dart, CapillarySchedules)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/shared/data/database.dart:
Could not analze DriftElementId(package:maismulher/features/hair_care/data/capillary_schedule_steps.dart, CapillaryScheduleSteps)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/shared/data/database.dart:
Could not analze DriftElementId(package:maismulher/features/hair_care/data/capillary_schedule_step_images.dart, CapillaryScheduleStepImages)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/shared/data/database.dart:
Could not analze DriftElementId(package:maismulher/features/storage/data/attached_images.dart, AttachedImages)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[WARNING] drift_dev on lib/features/shared/data/database.dart:
line 49, column 7 of package:maismulher/features/shared/data/database.dart: The referenced element could not be analyzed due to a bug in drift.
   ╷
49 │ class Database extends _$Database {
   │       ^^^^^^^^
   ╵
[WARNING] drift_dev on lib/features/shared/data/database.dart:
line 49, column 7 of package:maismulher/features/shared/data/database.dart: The referenced element could not be analyzed due to a bug in drift.
   ╷
49 │ class Database extends _$Database {
   │       ^^^^^^^^
   ╵
[WARNING] drift_dev on lib/features/shared/data/database.dart:
line 49, column 7 of package:maismulher/features/shared/data/database.dart: The referenced element could not be analyzed due to a bug in drift.
   ╷
49 │ class Database extends _$Database {
   │       ^^^^^^^^
   ╵
[WARNING] drift_dev on lib/features/shared/data/database.dart:
line 49, column 7 of package:maismulher/features/shared/data/database.dart: The referenced element could not be analyzed due to a bug in drift.
   ╷
49 │ class Database extends _$Database {
   │       ^^^^^^^^
   ╵
[WARNING] drift_dev on lib/features/shared/data/database.dart:
line 49, column 7 of package:maismulher/features/shared/data/database.dart: The referenced element could not be analyzed due to a bug in drift.
   ╷
49 │ class Database extends _$Database {
   │       ^^^^^^^^
   ╵
[WARNING] drift_dev on lib/features/shared/data/database.dart:
line 49, column 7 of package:maismulher/features/shared/data/database.dart: The referenced element could not be analyzed due to a bug in drift.
   ╷
49 │ class Database extends _$Database {
   │       ^^^^^^^^
   ╵
[WARNING] drift_dev on lib/features/shared/data/database.dart:
line 49, column 7 of package:maismulher/features/shared/data/database.dart: The referenced element could not be analyzed due to a bug in drift.
   ╷
49 │ class Database extends _$Database {
   │       ^^^^^^^^
   ╵
[WARNING] drift_dev on lib/features/storage/data/attached_images.dart:
Could not analyze DriftElementId(package:maismulher/features/storage/data/attached_images.dart, AttachedImages)
type 'InstanceCreationExpressionImpl' is not a subtype of type 'MethodInvocation' in type cast
[INFO] Running build completed, took 13.9s
[INFO] Caching finalized dependency graph completed, took 46ms
[INFO] Succeeded after 14.0s with 32 outputs (73 actions)
```

**Comment by simolus3** [comment]
`mixin SyncTable on Table` works, or if you want to use the new syntax I guess it's this:

```dart
abstract mixin class SyncTable implements Table {
  DateTimeColumn get updateDate =>
      dateTime().clientDefault(() => DateTime.now())();

  BoolColumn get synced => boolean().clientDefault(() => false)();

  BoolColumn get deleted => boolean().clientDefault(() => false)();
}

// usage:
@DataClassName('TodoCategory')
class TodoCategories extends Table with SyncTable {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get name => text()();
}
```

**Comment by JCKodel** [comment]
But override won't work. That's the whole point =\

I can't have this on a mixin:

```dart
@protected
@mustCallSuper
Future<void> onUpgrade(Database db, Migrator m, int from, int to) async {}
```

If you point up the place where non-decorated abstract classes are generated, I could develop some `@Ignore()` attribute and create a PR.

**Comment by simolus3** [comment]
I'm still not sure I understand, this works:

```dart
abstract mixin class SyncTable implements Table {
  DateTimeColumn get updateDate =>
      dateTime().clientDefault(() => DateTime.now())();

  BoolColumn get synced => boolean().clientDefault(() => false)();

  BoolColumn get deleted => boolean().clientDefault(() => false)();

  @protected
  @mustCallSuper
  Future<void> onUpgrade(Database db, Migrator m, int from, int to) async {}
}

@DataClassName('TodoCategory')
class TodoCategories extends Table with SyncTable {
  IntColumn get id => integer().autoIncrement()();
  TextColumn get name => text()();

  @override
  Future<void> onUpgrade(Database db, Migrator m, int from, int to) async {
    await super.onUpgrade(db, m, from, to);
    print('onUpgrade in todocategories');
  }
}
```

-------

Either way, treating abstract classes as tables is perhaps not that intuitive and complicates legitimate use-cases like yours. We could add a build option (say something like `dont_recognize_abstract_table_classes`) to disable that from happening and then enable that option by default in the next major drift update.

If you want to help with that, your contributions would be much appreciated! You can add options [in this file](https://github.com/simolus3/drift/blob/develop/drift_dev/lib/src/analysis/options.dart). The code that decides whether a Dart class is a drift table or not is here: https://github.com/simolus3/drift/blob/657ee5133faa3d04c226356231446700bc525bc9/drift_dev/lib/src/analysis/resolver/discover.dart#L181-L193

It has access to options via `_discoverStep._driver.options`, so it could be changed to return `false` if that option was enabled and the class is abstract.

**Comment by ZhulanovAA** [comment]
I seem to have the same issue, though in my case I don't have any errors, only undesirable tables suddenly appearing in my database, which in a sense makes it even more frustrating. If I understood correctly and this happens only because of using drift files, this issue was really hard to track in my case. 

Unnecessary tables appeared in generated files when I had added drift files, but as I had used them only for queries, there was no need for recreating the database or making any migrations. So I just continued using old database without realizing there were changes in the generated tables. And only now, after making some unrelated changes to some tables and recreating database file, I sudenly got undesirable tables from all my abstract classes, probably months after adding first drift file causing that issue.

I don't know how to fix it, but at the very least, shouldn't  generator give some warnings when it ended up generating tables that are not in my `@DriftDatabase(tables: [...])`? Or are there any reasons why such condition should be possible in the first place?

**Comment by simolus3** [comment]
I see, that frustration is understandable. I like the idea about adding a warning for this case, that has been implemented with b0946df3b78706c58376832e11b5ed476fdacaad.

**Comment by JCKodel** [comment]
This still doesn't solve the fact that the generated code won't compile.

I had to create an extension method to fill in the gaps... and that's a hack =\

If the class is abstract, it makes no sense at all to generate code for it (unless it is inherited by another non abstract class, then, in this case, the generator will merge the two declarations, as inheritance works this way)
